### PR TITLE
up-to-date version of Weyland-Yutani parser example

### PR DESCRIPTION
**Pull request by gvwilson**
This PR replaces #80 - once it is in `main`, @gvwilson will merge other branches created at DG and start writing the tutorial.

cc @nathan-stender

### CODE REVIEW

**Review by mdlu — CHANGES_REQUESTED** [review]
could you run `hatch run lint:fmt` on the branch? it looks like there are some linting issues causing tests to not pass

(this also helped me catch our README is out-of-date, thank you! https://github.com/Benchling-Open-Source/allotropy/pull/106)

File: src/allotropy/parser_factory.py [hunk 1]
```diff
@@ -45,13 +49,19 @@ class Vendor(Enum):
...
     Vendor.BECKMAN_VI_CELL_XR: ViCellXRParser,
+    Vendor.EXAMPLE_WEYLAND_YUTANI: ExampleWeylandYutaniParser,
     Vendor.NOVABIO_FLEX2: NovaBioFlexParser,
     Vendor.MOLDEV_SOFTMAX_PRO: SoftmaxproParser,
     Vendor.PERKIN_ELMER_ENVISION: PerkinElmerEnvisionParser,
     Vendor.ROCHE_CEDEX_BIOHT: RocheCedexBiohtParser,
 }
 
 
+VENDOR_NAME_LOOKUP: dict[str, Vendor] = {
```
Comment (mdlu) [inline src/allotropy/parser_factory.py:60]: (see note above)

File: src/allotropy/parsers/example_weyland_yutani/example_weyland_yutani_parser.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                    device_identifier=data.instrument.nickname,
+                ),
+                measurement_document=self._get_measurement_document(data),
+            )
+        )
+
+    def _get_measurement_time(self, data: Data) -> TDateTimeValue:
+        return self.get_date_time(
+            data.instrument.serial_number or "2023-12-31"
+        )  # FIXME
```
Comment (mdlu) [inline src/allotropy/parsers/example_weyland_yutani/example_weyland_yutani_parser.py:59]: we should certainly add this to our README/guidelines, but could you change comments like this to start with `TODO` and a short description?

File: src/allotropy/parsers/example_weyland_yutani/example_weyland_yutani_structure.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+
+@dataclass
+class Instrument:
+    serial_number: str
+    nickname: str
+
+    @staticmethod
+    def create() -> Instrument:
+        return Instrument(serial_number="", nickname="")  # FIXME
```
Comment (mdlu) [inline src/allotropy/parsers/example_weyland_yutani/example_weyland_yutani_structure.py:40]: same note as above about `TODO`s

**Review by mdlu — CHANGES_REQUESTED** [review]
everything else looks great to me, thanks so much for doing this!

File: scripts/csv_as_json.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import sys
+
+from allotropy.parser_factory import VENDOR_NAME_LOOKUP
+from allotropy.to_allotrope import allotrope_from_io
+
+
+def main():
+    """Main driver."""
+    args = parse_args()
+    vendor = VENDOR_NAME_LOOKUP[args.vendor]
```
[the remark this answers is not in the document]
Comment (gvwilson) [reply scripts/csv_as_json.py]: fixed as requested

File: src/allotropy/parser_factory.py (hunk 1 above)
Comment (gvwilson) [reply src/allotropy/parser_factory.py]: fixed as requested

File: src/allotropy/parsers/example_weyland_yutani/example_weyland_yutani_parser.py (hunk 2 above)
Comment (gvwilson) [reply src/allotropy/parsers/example_weyland_yutani/example_weyland_yutani_parser.py]: fixed as requested

File: src/allotropy/parsers/example_weyland_yutani/example_weyland_yutani_structure.py (hunk 3 above)
Comment (gvwilson) [reply src/allotropy/parsers/example_weyland_yutani/example_weyland_yutani_structure.py]: fixed as requested

**Comment by gvwilson** [comment]
Requested changes have been made, and local checks with `hatch run lint:fmt` and `hatch run test:test` pass. However, the GitHub action that does linting fails with complaints about type hints in both Perkin-Elmer and Weyland-Yutani, so the tests don't run - please let me know how to proceed.

**Comment by nathan-stender** [comment]
@gvwilson these look like typing issues, did you run `hatch run lint:typing`? (can also run `hatch run lint:all` - that's what the GH check does)

**Comment by gvwilson** [comment]
@nathan-stender @mdlu I've added an `assert_df` utility to check cases where a dataframe block might or might not be popped - alternative was to add a `required` Boolean flag to `pop_csv_block_as_df`, and I'm happy to back out the most recent commit and do that instead to make type linting happy if you'd prefer.

**Comment by nathan-stender** [comment]
assert_df is good! Matches out existing patterns.

**Comment by gvwilson** [comment]
Thanks - we now have a clean build.

**Review by mdlu — APPROVED** [review]

**Comment by gvwilson** [comment]
Thanks for the approval @mdlu - apologies for taking so long to get it into shape. Once it's merged I'll close outstanding PRs that it subsumes and start work on the tutorial.

**Comment by mdlu** [comment]
closed in favor of https://github.com/Benchling-Open-Source/allotropy/pull/110

**Pull request closed without merging**